Frame.cc

```diff
@@ -1153,3 +1186,188 @@ bool Frame::GetSpectralData(std::vector<float>& data, casacore::SubImage<float>&
+    std::vector<int> requested_stats(config.stats_types().begin(), config.stats_types().end());
+    // initialize the size of statistical results
+    std::vector<std::vector<double>> results(stats_size);
+    for (int i = 0; i < stats_size; ++i) {
+        results[i].resize(profile_size);
+    }
+    // get region state for this process
+    RegionState region_state = region->GetRegionState();
+    // get statistical profile data
+    int start = 0;
+    int count, end;
+    casacore::SubImage<float> sub_image;
+    while (start < profile_size) {
+        // start the timer
+        auto tStart = std::chrono::high_resolution_clock::now();
+        // check if frontend queries changed, if so, terminate this loop process
+        if (!_connected) {
+            std::cerr << "[Region " << region_id << "] closing image, exit zprofile (statistics) before complete" << std::endl;
+            return false;
+        }
+        if (_region_states.count(region_id) && _region_states[region_id] != region_state) {
+            std::cerr << "[Region " << region_id << "] region state changed, exit zprofile (statistics) before complete" << std::endl;
+            return false;
+        }
+        if (_region_configs.count(region_id) && !_region_configs[region_id].IsAmong(profile_index, requested_stats)) {
+            std::cerr << "[Region " << region_id << "] region requirement changed, exit zprofile (statistics) before complete" << std::endl;
+            return false;
+        }
+        end = (start + delta_channels > profile_size ? profile_size - 1 : start + delta_channels - 1);
+        count = end - start + 1;
+        GetRegionSubImage(region_id, sub_image, profile_stokes, ChannelRange(start, end));
+        std::vector<std::vector<double>> buffer;
+        if (region->GetSpectralProfileData(buffer, profile_index, sub_image)) {
+            for (int j = 0; j < stats_size; ++j) {
+                memcpy(&results[j][start], &buffer[j][0], count * sizeof(double));
+            }
+        } else {
+            std::cerr << "Can not get zprofile (statistics), region id: " << region_id
+                      << ", channel range: ["<< start << "," << end << "]" << std::endl;
+            return false;
+        }
+        start += count;
+        // get the time elapse for this step
+        auto tEnd = std::chrono::high_resolution_clock::now();
+        auto dt = std::chrono::duration<double, std::milli>(tEnd - tStart).count();
+        // adjust the increment of channels according to the time elapse
+        delta_channels *= dt_target / dt;
+        if (delta_channels < 1) {
+            delta_channels = 1;
+        }
+        if (delta_channels > profile_size) {
+            delta_channels = profile_size;
+        }
+    }
+    stats_values = std::move(results);
+    return true;
+}
+
+bool Frame::CalcRegionSpectralData(std::map<CARTA::StatsType, std::vector<double>>& stats_values, int stokes, int region_id,
```

Comment:
I strongly feel that this code should remain inside the HDF5 loader -- we have consistently factored format-specific code out of the frame object. The frame object should not be aware of HDF5-specific implementation details -- it should only call loader functions to check if the loader has a particular feature or can produce some dataset.

Reply:
there might be a need for some of the logic and final stats combining to be placed in the Frame, especially since we want to return _partial_ results to the frontend. But I don't think just moving the entire function to `Frame` is the right solution.

Reply:
@markccchiang please keep in mind that the end goal is to also send partial results to the frontend at regular intervals, whether those partial results are from the original dataset or the swizzled data.

Reply:
I see. I will move it back to the HDF5 loader. Then the loader will be responsible for determining whether to interrupt the stats calculation.

Reply:
@confluence can you give some thought as to how we can return partial results from the loader while it is still in progress? The messages to the frontend will get sent from `Session`. Not sure if passing a callback into the frame and loader functions would be the most elegant solution, but it might work

Reply:
I think a callback is the most sensible solution -- it seems like a good place to start.

Reply:
I wonder if a callback could also be used to allow the loader to check whether the region is still valid

Reply:
BTW, sending partial results from the original dataset is no problem. But it looks unlikely for swizzled data?

Reply:
Well, they'd be partial in a different direction. :) You'd have to recalculate the mean, RMS, etc., after every pass, and they'd get more accurate as you read in more slices, until eventually you'd have the exact values.